### DISCUSSION TITLE
How to access the selected fields before field resolution?

### DISCUSSION THREAD

**Discussion by benmccann** [graphql-java-kickstart/graphql-java-tools#397] [Q&A] [answered]
How do we get the list of user-selected fields? [The examples in the docs](https://www.graphql-java-kickstart.com/tools/schema-definition/) all show returning the full object, but that's often overfetching. Is there a way to select just the fields that the user has asked for?

**Comment by benmccann** [comment]
Answering my own question, you can put the `DataFetchingEnvironment` as the last parameter to the resolver method and then call `env.getSelectionSet`. I'll leave this open though because I think it might be useful to document

**Comment by vojtapol** [comment, marked as answer]
The solution you are describing is the [field look-ahead strategy](https://www.graphql-java.com/blog/deep-dive-data-fetcher-results/) and it can be quite powerful for optimizing fetching. In general, I would not recommend it though. One big disadvantage is that you cannot look-ahead into fragments of subtypes because at the time of looking-ahead GraphQL Java still does not know what type will the fragment resolve into.

To avoid over-fetching I would instead recommend fetching just the minimal "skeleton" objects and resolving all additional fields using methods in `GraphQLResolver<MyType>`. If you then combine it with the [GraphQL Java DataLoader](https://www.graphql-java.com/documentation/v12/batching/) concept you will have a very flexible and fast architecture with little to no overfetching.
